### PR TITLE
Add Frame Time Recording

### PR DESCRIPTION
**Pull request by BreadFish64**
Frame times are recorded until the game stops or the user unchecks this option:
image
Then the data is dumped to a .csv file in the log directory and can be used to make fancy graphs like this:
image

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4636)

### CODE REVIEW

File: src/core/settings.h [hunk 1]
```diff
@@ -175,6 +175,7 @@ struct Values {
     // Debugging
     bool use_gdbstub;
     u16 gdbstub_port;
+    std::atomic_bool record_frame_times;
```
Comment (zhaowenlan1779) [inline src/core/settings.h]: I think this shouldn't be a part of the settings; I think it's more suitable for being a temporary state in PerfStats.

File: src/core/perf_stats.cpp [hunk 2]
```diff
@@ -73,6 +86,31 @@ double PerfStats::GetLastFrameTimeScale() {
...
+            auto in_time_t = std::chrono::system_clock::to_time_t(now);
+            std::stringstream path;
+            path << FileUtil::GetUserPath(FileUtil::UserPath::LogDir) << game_title << " "
+                 << std::put_time(std::localtime(&in_time_t), "%F %H-%M-%S") << ".csv";
+
+            FileUtil::WriteStringToFile(true, data, path.str().data());
+        },
+        frame_data->str());
+
+    write_thread.detach();
```
Comment (zhaowenlan1779) [inline src/core/perf_stats.cpp]: I think you should use Common::DetachedTasks for this purpose

File: src/citra_qt/main.ui [hunk 3]
```diff
@@ -155,6 +155,7 @@
      <addaction name="action_Advance_Frame"/>
     </widget>
     <addaction name="menu_Movie"/>
+    <addaction name="action_Record_Frame_Times"/>
```
Comment (zhaowenlan1779) [inline src/citra_qt/main.ui]: regarding UI designs, I'd argue that it would be better to put the submenus together, above the actions.

File: src/citra_qt/main.ui [hunk 4]
```diff
@@ -45,7 +45,7 @@
      <x>0</x>
      <y>0</y>
      <width>1081</width>
-     <height>21</height>
+     <height>26</height>
```
Comment (zhaowenlan1779) [inline src/citra_qt/main.ui]: why is this changed?

File: src/core/core.cpp [hunk 5]
```diff
@@ -279,6 +281,7 @@ void System::RegisterSoftwareKeyboard(std::shared_ptr<Frontend::SoftwareKeyboard
 }
 
 void System::Shutdown() {
+    perf_stats->FlushFrameData();
```
Comment (tgsm) [inline src/core/core.cpp]: This causes a crash on exit if dumping frame times is turned off.

File: src/core/perf_stats.cpp [hunk 6]
```diff
@@ -73,6 +86,31 @@ double PerfStats::GetLastFrameTimeScale() {
...
+void PerfStats::FlushFrameData() {
+    std::string game_title;
+    system.GetAppLoader().ReadTitle(game_title);
+    std::thread write_thread(
+        [game_title](std::string data) { // file path is "log/<game title> <date> <time stamp>.csv"
+            auto now = std::chrono::system_clock::now();
+            auto in_time_t = std::chrono::system_clock::to_time_t(now);
+            std::stringstream path;
+            path << FileUtil::GetUserPath(FileUtil::UserPath::LogDir) << game_title << " "
+                 << std::put_time(std::localtime(&in_time_t), "%F %H-%M-%S") << ".csv";
```
Comment (zhaowenlan1779) [inline src/core/perf_stats.cpp]: I personally like fmt::format better than the std stream formatting. just a personal style opinion though.

File: src/citra_qt/main.ui (hunk 4 above)
Comment (BreadFish64) [reply src/citra_qt/main.ui]: I blame qt designer

File: src/core/perf_stats.cpp (hunk 6 above)
Comment (BreadFish64) [reply src/core/perf_stats.cpp]: Are you referring to the `"%F %H-%M-%S"`? That is part of the the puttime function so I don't think I can change it

File: src/core/settings.h (hunk 1 above)
Comment (BreadFish64) [reply src/core/settings.h]: Like not saving it between runs? I don't see the advantage to that

File: src/core/perf_stats.cpp (hunk 6 above)
Comment (zhaowenlan1779) [reply src/core/perf_stats.cpp]: nope, I mean the stringstream there. it can be replaced with fmt

File: src/core/settings.h (hunk 1 above)
Comment (zhaowenlan1779) [reply src/core/settings.h]: I'm not saying you shouldn't save it (though I doubt the necessity of this), but even if you do, you shouldn't use the settings as a state of PerfStats. (and if you look carefully all other fields here are not atomic)

Instead make this just a `bool` and add an `atomic_bool` to `PerfStats`, and initialize that atomic bool from the settings on load and save it on exit.

**Comment by BreadFish64** [comment]
review comments addressed and bug squashed

**Comment by BreadFish64** [comment]
refactored to not use unique_ptr

**Comment by JayFoxRox** [comment]
I'm currently strongly against this. It appears totally unnecessary to me.
The feature is also very underdocumented: how would I know the output will land in the log, just from that menu option?

Profiling will be done differently, using more powerful tools.
Apparently microprofile is also still part of Citra, so I assume this functionality exists already (to some extend).

There's just a ton of unanswered questions, especially considering this PR has been here for 14 days already:

- Why would anyone use this feature? Fancy graphs for... what?
- Why can't it be solved using the existing tools like microprofile?
- Why does it have to be solved in Citra, not a third-party solution which most people will have to use anyway?
- Why is this considered a "Tool" and not specifically marked as a debug feature?
- ...

**Comment by BreadFish64** [comment]
>There's just a ton of unanswered questions

Nobody asked any, now that there's questions I can answer them.

>Why can't it be solved using the existing tools like microprofile?

I wasn't aware that microprofile had this capability because it's poorly documented in the application; searching "microprofile" on google gave irrelevant results and nobody told me for the 14 days this was open.  I don't see an option to actually save the data microprofile outputs but maybe that would be a better approach.

>Why does it have to be solved in Citra, not a third-party solution which most people will have to use anyway?

Most third party applications used for measuring frame data such as MSI afterburner don't work properly in conjunction with Citra as far as I can remember.

>Why is this considered a "Tool" and not specifically marked as a debug feature?

Because I saw reasons for non-developers to use it. It was mentioned on the yuzu discord that it might be useful for youtubers to get hard performance. (A single checkbox is also much less intimidating to these users than a separate UI full of seemingly redundant options, some of which have a non-trivial performance if enabled accidentally.)

>...

If you have other feedback then you can say it.

**Comment by GXTX** [comment]
If the feature is only going to be used in very specific instances `YouTubers gathering hard performance data`, but even then it's a very very small subset of users that would ever touch it, shouldn't this be a debug only feature? Or is this something that needs to be mainlined in Citra instead of pointing the user in the right direction with 3rd party tools specifically created for these tasks?

Seems to be like this is just feature creep in action.

**Comment by BreadFish64** [comment]
I gave one possible use for it, personally I was using it to compare performance data with different compiler optimizations in Excel

**Comment by wwylele** [comment]
I'll put one of my concern expressed on discord here to remind that this not resolved yet:

> IMO feature creep isn't a real issue if it doesn't detract from user experience.

Feature creep is more of a dev issue rather than a user issue that might hinder the development from making essential process.

**Comment by BreadFish64** [comment]
I'm going to close this. It seems like the better approach would be to use microprofile to save the data but I didn't want to spend a lot time on this in the first place.
